anaconda_project/plugins.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------------
+# Copyright © 2016, Continuum Analytics, Inc. All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+"""Plugins assets."""
+import os
+from os.path import join
+from anaconda_project.project_commands import (_ArgsTransformer, ProjectCommand, _http_specs)
+
+from entrypoints import get_group_named
+
+
+def get_all(entry_points_group='anaconda_project.plugins'):
+    """Return all the plugins registered."""
+    return {name: plugin.load() for name, plugin in sorted(get_group_named(entry_points_group).items())}
+
+
+class ArgsTrasformerTemplate(_ArgsTransformer):
+    """Template class for plugins args trasformers.
+    Plugins args transformers should subclass it and redefine add_class
+    to implement custom arguments mapping.
+    """
+
+    def __init__(self, command):
+        """Construct an ArgTransformer for the given command.
+
+        Args:
+            command (ProjectCommand): command that maps to the ArgsTransformer
+        """
+        # TODO: document specs
+        # TODO: decide the best way to let users define specs for their
+        #       transformers! Most likely specs should go on the command
+        #       itself?
+        super(ArgsTrasformerTemplate, self).__init__(_http_specs)
+        self.command = command
+
+    def add_args(self, results, args):
+        """Overwrite this method to add custom arguments transformation.
+        It should forwarding the arguments that are custom to the
+        specific command served by this trasformer (
+            i.e., '--anaconda-project-host' --> 'host'
+        )
+
+        Inputs:
+            - results [list(tuples)]: list of 2 element tuples the following:
+                           * option (str): name of the option
+                            * values (lst(str)): list of the values passed
+                                        for the option
+            - args [list]: list of the args already passed in
+
+        Returns:
+            (list) list of the transformed args (that should include args)
+        """
+        raise RuntimeError("not implemented")  # pragma: no cover
+
+
+class CommandTemplate(ProjectCommand):
```

Comment:
Allowing the plugin to subtype ProjectCommand exposes a pretty large customization surface, it will be hard to know what it's OK to change in the anaconda_project core if a plugin could have overridden any part of ProjectCommand.

It could be safer (at least to start) to restrict plugins to more specific function hooks, like a transform_args function... but of course that means plugins can't do as much stuff.